File: tools/lib/snippets.js
// 보안 스니펫 본문을 소스에서 떼어내는 부분. `check-snippets.js`(구문 검사)와
// `check-snippet-golden.js`(구워진 결과 잠그기)가 같은 것을 봐야 하므로 여기 모아 둔다.
//
// **앵커 목록을 손으로 관리하지 않는다.** 호출 형태로 자동 수집한다 - 목록 방식은 새
// 스니펫을 추가할 때 등록을 잊으면 조용히 검사에서 빠지고, 그게 정확히 이 도구들이 막으려던
// 실패 방식이다.
const fs = require("fs");
const path = require("path");

// 인자에 스니펫 본문이 오는 호출들.
//
// **여기에 안 적힌 이름으로 본문을 넘기면 그 스니펫은 조용히 검사 밖이 된다.** `InstallSnippet`을
// 도입하면서 실제로 46개가 43개로 줄었고, 줄어든 셋이 클릭 래퍼였다 - 이 저장소에서 제일 뜨거운
// 경로다. 개수가 줄면 그것부터 의심할 것.
const CALLS = /\b(SecureHandlerExecute|SecureHandlerWrapScript|SetAttribute|InstallSnippet)\s*\(/g;

// `local X_SNIPPET = ... [[본문]] ... or ""` 꼴. 스니펫이 조각으로 나뉠 때 쓰는 관용구라
// (DEBUG 빌드에서만 들어가는 부분) 참조를 풀어서 결합된 형태도 같이 본다.
const SNIPPET_LOCAL = /\blocal\s+([A-Za-z_][\w]*_SNIPPET)\s*=/g;

// **조각이 전부 DEBUG 전용인 것은 아니다.** 한 본문을 둘 이상이 나눠 쓰려고 뽑아둔 조각은
// 언제나 들어간다. 그런 것까지 "빠진 빌드"로 세면 골든이 존재하지 않는 형상을 기록하고, 그
// 형상에서 뽑아낸 줄들은 검사를 벗어난다 - 조각으로 뽑는 행위 자체가 검사에서 빠지는 길이 된다.
//
// 가르는 표시는 **선언이 DEBUG에 걸려 있는가**다. 뒤의 `or ""`가 아니라 앞의 조건을 보는
// 이유는 그게 실제 게이트라서다 - 빠지는 빌드가 있다는 것을 정하는 것은 그 조건이지, 안 걸렸을
// 때 무엇으로 떨어지느냐가 아니다.
const DEBUG_GATED = /\bDEBUG\b/;

/** 위치 i에서 시작하는 긴 문자열을 뗀다. 아니면 null. */
function readLongString(src, i) {
    if (src[i] !== "[") return null;
    let j = i + 1;
    while (src[j] === "=") j++;
    if (src[j] !== "[") return null;
    const level = j - i - 1;
    let body = j + 1;
    // 여는 대괄호 바로 뒤의 줄바꿈 하나는 본문에 안 들어간다(Lua 규칙).
    if (src[body] === "\r") body++;
    if (src[body] === "\n") body++;
    const close = "]" + "=".repeat(level) + "]";
    const end = src.indexOf(close, body);
    if (end < 0) return null;
    return { body: src.slice(body, end), next: end + close.length };
}

/**
 * 호출의 여는 괄호부터 짝이 맞는 닫는 괄호까지 훑으며 인자에 실린 본문들을 모은다.
 * 주석·따옴표 문자열·중첩 괄호를 건너뛰는 작은 스캐너다. 정규식으로는 긴 문자열 안의
 * 괄호와 진짜 괄호를 못 가른다.
 *
 * **인자 하나가 본문 하나다.** `..`로 이어진 조각은 한 본문으로 합치고, 인자를 가르는
 * 쉼표에서는 끊는다. `SecureHandlerWrapScript`는 preBody와 postBody를 따로 받으므로
 * 안 끊으면 둘을 이어붙인 엉뚱한 소스를 검사하게 된다.
 */
function collectBodies(src, openParen, snippetLocals) {
    const bodies = [];
    let parts = [];
    let depth = 1;
    let i = openParen + 1;

    const flush = () => {
        if (parts.length) bodies.push(parts);
        parts = [];
    };

    while (i < src.length && depth > 0) {
        const two = src.substr(i, 2);

        if (two === "--") {
            const long = readLongString(src, i + 2);
            if (long) { i = long.next; continue; }
            const nl = src.indexOf("\n", i);
            i = nl < 0 ? src.length : nl + 1;
            continue;
        }

        const long = readLongString(src, i);
        if (long) {
            parts.push({ literal: long.body });
            i = long.next;
            continue;
        }

        const c = src[i];
        if (c === '"' || c === "'") {
            i++;
            while (i < src.length && src[i] !== c) {
                i += src[i] === "\\" ? 2 : 1;
            }
            i++;
            continue;
        }

        if (c === "(") { depth++; i++; continue; }
        if (c === ")") { depth--; i++; continue; }
        if (c === "," && depth === 1) { flush(); i++; continue; }

        const ident = /^[A-Za-z_][\w]*/.exec(src.slice(i));
        if (ident) {
            if (snippetLocals.has(ident[0])) parts.push({ ref: ident[0] });
            i += ident[0].length;
            continue;
        }

        i++;
    }

    flush();
    return bodies;
}

/**
 * 문자열과 주석을 같은 길이의 공백으로 덮는다. **길이와 줄바꿈을 보존하므로** 이 결과에서
 * 찾은 위치를 원본에 그대로 쓸 수 있다.
 *
 * 쓰이는 데가 둘이다. 호출을 찾을 때 - 주석에 적힌 `SecureHandlerWrapScript(...)`를 진짜
 * 호출로 오해하면 안 된다. 그리고 5.1 토큰 검사에서 - 메시지 안의 색 코드("|cffff6666")를
 * 비트 or로 읽으면 안 된다.
 */
/**
 * Blanks comments and string contents to spaces, keeping every other position and every newline.
 *
 * `commentStarts`, when given an array, collects the index each comment opens at. That is the one
 * thing the mask alone cannot answer afterwards: a blanked run could have been a comment or a
 * string, and telling a trailing comment from a `"--"` inside a literal needs to know which.
 */
function blankNonCode(src, commentStarts) {
    const out = Array.from(src);
    const blank = (from, to) => {
        for (let k = from; k < to && k < out.length; k++) {
            if (out[k] !== "\n" && out[k] !== "\r") out[k] = " ";
        }
    };

    let i = 0;
    while (i < src.length) {
        if (src.substr(i, 2) === "--") {
            if (commentStarts) commentStarts.push(i);
            const long = readLongString(src, i + 2);
            const end = long ? long.next : (src.indexOf("\n", i) < 0 ? src.length : src.indexOf("\n", i));
            blank(i, end);
            i = end;
            continue;
        }
        const long = readLongString(src, i);
        if (long) { blank(i, long.next); i = long.next; continue; }
        const c = src[i];
        if (c === '"' || c === "'") {
            const start = i;
            i++;
            while (i < src.length && src[i] !== c) i += src[i] === "\\" ? 2 : 1;
            i++;
            blank(start, i);
            continue;
        }
        i++;
    }
    return out.join("");
}

/**
 * Collects the `local X_SNIPPET = [[body]]` pieces by name. An argument carrying one of those
 * names is only a name, so this is what turns it back into a body -- and it is also where a tool
 * that wants one piece on its own gets it.
 *
 * `code` is the same-length mask `blankNonCode` produces. Pass it if you already have one.
 */
function collectSnippetLocals(src, code) {
    code = code || blankNonCode(src);

    const snippetLocals = new Map();
    SNIPPET_LOCAL.lastIndex = 0;
    let m;
    while ((m = SNIPPET_LOCAL.exec(code))) {
        const long = (() => {
            for (let i = m.index + m[0].length; i < src.length; i++) {
                const s = readLongString(src, i);
                if (s) return { ...s, start: i };
                if (src[i] === "\n" && src.slice(m.index, i).includes(";")) return null;
            }
            return null;
        })();
        if (long) {
            // 본문 앞, 즉 `local X_SNIPPET =`와 여는 괄호 사이가 게이트가 적히는 자리다.
            const head = src.slice(m.index, long.start);
            snippetLocals.set(m[1], { body: long.body, gated: DEBUG_GATED.test(head) });
        }
    }
    return snippetLocals;
}

/**
 * `Debind/`의 모든 스니펫 본문을 순서대로 넘긴다. 콜백 인자:
 *   { file, line, call, label, body }
 *
 * `label`은 DEBUG 갈래를 가르는 이름이다. 조각 참조가 섞인 본문은 **양쪽 다** 넘긴다 -
 * 릴리스 빌드에서는 그 자리가 빈 문자열이 되어 결합 지점이 달라지므로, 한쪽에서만 나는
 * 문제가 실제로 있을 수 있다.
 */
function forEachSnippet(srcDir, cb) {
    const files = fs.readdirSync(srcDir).filter((f) => f.endsWith(".lua")).sort();

    for (const file of files) {
        const src = fs.readFileSync(path.join(srcDir, file), "utf8");
        // 위치를 보존하는 마스크. 여기서 찾은 인덱스를 원본에 그대로 쓴다.
        const code = blankNonCode(src);

        // 1. 조각으로 쓰이는 스니펫 지역변수를 먼저 모은다.
        const snippetLocals = collectSnippetLocals(src, code);

        // 2. 호출마다 본문을 모아 넘긴다.
        let m;
        CALLS.lastIndex = 0;
        while ((m = CALLS.exec(code))) {
            const openParen = m.index + m[0].length - 1;
            const line = src.slice(0, m.index).split("\n").length;

            for (const parts of collectBodies(src, openParen, snippetLocals)) {
                if (!parts.some((p) => p.literal !== undefined)) continue;

                // 게이트가 붙은 참조가 하나라도 있을 때만 두 형상이 존재한다. 무조건 들어가는
                // 조각은 어느 쪽에서도 본문 그대로다.
                const resolve = (p) => (p.ref ? snippetLocals.get(p.ref).body : p.literal);
                const hasGated = parts.some((p) => p.ref !== undefined && snippetLocals.get(p.ref).gated);
                const variants = hasGated
                    ? [["DEBUG=on", resolve],
                       ["DEBUG=off", (p) => (p.ref && snippetLocals.get(p.ref).gated ? "" : resolve(p))]]
                    : [["", resolve]];

                for (const [label, pick] of variants) {
                    cb({ file, line, call: m[1], label, body: parts.map(pick).join("") });
                }
            }
        }
    }
}

module.exports = {
    readLongString, blankNonCode, collectBodies, collectSnippetLocals, forEachSnippet,
    CALLS, SNIPPET_LOCAL,
};

File: Debind/Snippets.lua
local _, DebindPrivate = ...;
local Constants = DebindPrivate.Constants;

-- `string.format`, not the `format` global. The same one in the game, but this file is also run
-- by `tools/` under a shim that has no WoW globals -- and until a probe with a live form appeared
-- in a baked body, no call site here was ever reached from the tools to say so.
local format = string.format;

--- Marks which lines *begin* in code, as opposed to beginning inside something that started on
--- an earlier line -- a long string or a long comment.
---
--- `StripSnippetComments` needs exactly this and nothing more. Deciding line by line, it cannot
--- otherwise tell a comment from a line of long-string content that happens to open with `--`,
--- and emptying the latter corrupts the body silently. Scanning for `[[` in the raw text is not
--- a substitute: `-- [[` is an ordinary comment by Lua's rule, since a long bracket only opens
--- one when it follows `--` immediately, and a scan without context reads it as a string that
--- never closes.
---
--- Short strings are tracked too, so a `"--"` inside one cannot be mistaken for a comment.
--- An unterminated one gives up at the newline rather than swallowing the rest of the body.
local function codeLineStarts(str)
	local starts = { [1] = true };
	local line = 1;
	local i, n = 1, #str;

	local function crossNewline(isCode)
		line = line + 1;
		starts[line] = isCode;
	end

	--- Skips a long bracket from its opener, marking every line inside it as not code.
	--- An unclosed one runs to the end, which is what the game would see too.
	local function skipLongBracket(from, eqs)
		local close = "]" .. eqs .. "]";
		local ce = str:find(close, from, true);
		local stop = ce and (ce + #close) or (n + 1);

		for k = from, stop - 1 do
			if (str:sub(k, k) == "\n") then crossNewline(false); end
		end

		return stop;
	end

	while (i <= n) do
		local c = str:sub(i, i);

		if (c == "\n") then
			crossNewline(true);
			i = i + 1;

		elseif (c == "-" and str:sub(i + 1, i + 1) == "-") then
			local eqs = str:match("^%[(=*)%[", i + 2);
			if (eqs) then
				i = skipLongBracket(i + 2, eqs);
			else
				-- To end of line. The newline itself is left for the next round so the line
				-- after a comment is still marked as starting in code.
				i = str:find("\n", i, true) or (n + 1);
			end

		elseif (c == "[") then
			local eqs = str:match("^%[(=*)%[", i);
			i = eqs and skipLongBracket(i, eqs) or (i + 1);

		elseif (c == '"' or c == "'") then
			i = i + 1;
			while (i <= n) do
				local d = str:sub(i, i);
				if (d == "\\") then
					i = i + 2;
				elseif (d == c) then
					i = i + 1;
					break;
				elseif (d == "\n") then
					break;
				else
					i = i + 1;
				end
			end

		else
			i = i + 1;
		end
	end

	return starts;
end

--- Drops comment text from a snippet body. **Line-leading comments only** -- a trailing
--- `x = a -- why` would need to know whether the `--` is inside a string literal, and getting
--- that wrong corrupts a body silently. Every comment in these bodies already starts its line.
---
--- The line itself stays, emptied. Dropping it would shift every line after it, and the line
--- numbers a snippet reports on error are the only handle there is for locating a fault inside
--- one. Only the text goes.
---
--- Stripping runs **before** substitution, which is what keeps two hazards from existing at
--- all: a comment mentioning a token would otherwise be substituted, and a replacement longer
--- than one line would otherwise spill out of the comment it landed in and become code.
---
--- Exposed rather than local because `tools/check-snippet-golden.js` calls this exact function
--- through fengari. That tool locks what the game receives, so a second copy of this rule in
--- JavaScript would be a copy that can drift -- and it would drift silently, since the tool
--- would then be guarding something other than what gets baked.
function DebindPrivate.StripSnippetComments(str)
	local starts = codeLineStarts(str);
	local out = {};
	local pos, len, no = 1, #str, 1;

	while (pos <= len) do
		local nl = str:find("\n", pos, true);
		local line = str:sub(pos, nl and nl - 1 or len);
		local rest = starts[no] and line:match("^[ \t]*%-%-(.*)$");

		if (rest) then
			-- A long comment opening the line would have its `--[[` emptied while the body it
			-- opened stays, leaving that text and a stray `]]` behind as code. Only the opener
			-- is on this line, so there is nothing to check further in; refuse it outright.
			assert(not rest:match("^%[=*%["),
				"long-bracket comment starting a snippet line - write it as plain `--` lines");
			out[#out + 1] = "";
		else
			out[#out + 1] = line;
		end

		no = no + 1;
		if (not nl) then break; end
		pos = nl + 1;
	end

	return table.concat(out, "\n");
end

--- Snippet bodies are strings, so anything the restricted environment cannot reach has to be
--- folded in before the body is handed to `SetAttribute`. This is where that folding lives.
---
--- It sits here rather than in `SecureBindings.lua` because `UpdateBindings.lua` generates
--- snippet text too, and both have to fold the same way from the same table.
---
--- **반드시 값 하나만 돌려준다.** `gsub`은 (문자열, 치환횟수)를 주는데, 그걸 그대로 흘리면
--- 마지막 인자 자리에서 남는 값이 다음 매개변수로 들어간다. `SetAttribute(name, body)`는
--- 남는 인자를 무시해서 표가 안 났지만, `SecureHandlerWrapScript(f, script, header, preBody,
--- postBody)`에서는 치환 횟수가 postBody가 되어 "Invalid post-handler body"로 터진다.
--- 그리고 그 오류는 **파일의 나머지를 통째로 중단시킨다** - 뒤따르는 속성들이 전부 정의되지
--- 않은 채로 게임이 계속 돌아서, 증상이 엉뚱한 곳(FrameRegistry의 OnEnter 래핑)에서 난다.
--- What each `PROBE.<name>(...)` becomes in a shipped build.
---
--- The point of writing them as probes at all is that they are **the plain call and nothing
--- else** once baked. A body carrying `PROBE.UnitExists(unit)` produces the same bytes as one
--- that always said `UnitExists(unit)`; `tools/snippet-golden.txt` is what holds that to it.
---
--- A name mapping to `false` disappears entirely, which is how a probe that only exists to report
--- something outward costs a real user nothing at all -- not a check that fails, not a table
--- lookup that misses, no line.
DebindPrivate.SNIPPET_PROBES_LIVE = {
	UnitExists = "UnitExists(%s)",
	PlayerCanAssist = "PlayerCanAssist(%s)",
	PlayerCanAttack = "PlayerCanAttack(%s)",
	PlayerIsChanneling = "PlayerIsChanneling(%s)",
	SecureCmdOptionParse = "SecureCmdOptionParse(%s)",

	-- Reporting only. Nothing is computed from it, so there is nothing to keep.
	Winner = false,

	-- Injection only. The click path measures its own axes now, and a test that wants to say
	-- "you are in combat" has to reach the value between the measurement and the comparison --
	-- writing into `States` does not hold, because the click measures again rather than reading
	-- it. Absent here, so a real user's snippet has no table to miss and no branch to fail.
	MockState = false,
};

--- Replaces the `PROBE.<name>(args)` tokens.
---
--- Arguments are taken as far as the first `)`, which is all these ever need and keeps the
--- substitution a regular one. A nested call would simply not match, and the leftover token is
--- caught below rather than compiled into a snippet that fails somewhere else.
local function applyProbes(str, table_)
	local result = str:gsub("PROBE%.([_A-Za-z0-9]+)%(([^()]*)%)", function(name, args)
		local form = table_[name];
		assert(form ~= nil, "unknown probe: " .. name);
		if (form == false) then
			return "";
		end
		return format(form, args);
	end);

	assert(not result:find("PROBE%.", 1, false),
		"a PROBE token survived substitution - nested parentheses in its arguments?");

	return result;
end

-- Exposed for `tools/check-snippet-golden.js`, which bakes every body with the live table to see
-- that a probe changes nothing. A second copy of this rule in JavaScript could drift, and a
-- drifted one would be guarding something other than what gets baked.
DebindPrivate.applyProbesForTools = applyProbes;

-- Raw bodies kept alongside the thing that installs them, so a body can be baked again later.
--
-- Bodies are baked while this addon loads, which is before anything that might want a probe
-- turned on has loaded at all -- a dependency loads after what it depends on, and there is no
-- ordering that puts it first. So the choice cannot be made once at bake time; the bodies have to
-- remain re-bakeable.
--
-- That is also the strongest form of the switch. Turning probes off is not "read an empty table"
-- but a rebuild with the live table, which produces the bytes a real user runs -- the same test
-- can be run against code that has no probes in it at all.
local rebakeable = {};

--- Bakes each body and hands them all to `install`, keeping them so it can be done again.
---
--- Takes several because `SecureHandlerWrapScript` has two -- a pre and a post -- and both are
--- snippets. `install` receives the baked texts in order. Everything a rebake needs is captured
--- in that closure, so the registry does not need to know what kind of destination it is: an
--- attribute, a wrapped script, or something that does not exist yet.
function DebindPrivate.InstallSnippet(install, ...)
	local bodies = { ... };
	rebakeable[#rebakeable + 1] = { install = install, bodies = bodies };

	local baked = {};
	for i = 1, #bodies do
		baked[i] = DebindPrivate.BakeSnippet(bodies[i]);
	end
	install(unpack(baked));
end

--- Bakes every registered body again with whatever table is current.
---
--- Out of combat only, which costs nothing: the environment this exists to drive is one where the
--- client is never actually in combat.
function DebindPrivate.RebakeSnippets()
	if (InCombatLockdown()) then
		return false, "in combat";
	end

	for i = 1, #rebakeable do
		local entry = rebakeable[i];
		local baked = {};
		for j = 1, #entry.bodies do
			baked[j] = DebindPrivate.BakeSnippet(entry.bodies[j]);
		end
		entry.install(unpack(baked));
	end

	return true;
end

function DebindPrivate.BakeSnippet(str)
	local probes = (DebindPrivate.SnippetProbes and DebindPrivate.SnippetProbes.expand)
		or DebindPrivate.SNIPPET_PROBES_LIVE;

	local result = applyProbes(DebindPrivate.StripSnippetComments(str), probes)
		:gsub("CONSTANTS%.([_A-Za-z0-9]+)", function(m)
		local value = Constants[m];
		assert(value ~= nil, m);
		if (type(value) == "string") then
			return format("%q", value);
		else
			return tostring(value);
		end
	end);
	return result;
end
